boxtree/tree.py

```diff
@@ -111,9 +114,109 @@ class box_flags_enum(Enum):  # noqa
 # }}}
 
 
+# {{{ tree of boxes
+
+@dataclass
+class TreeOfBoxes:
+    """A quad/octree tree of abstract boxes. Lightweight trees handled with
+    :mod:`numpy`, intended for mesh adaptivity.
+
+    .. attribute:: dimensions
+
+    .. attribute:: nlevels
+
+    .. attribute:: nboxes
+
+    .. attribute:: root_extent
+
+        (Scalar) extent of the root box.
+
+    .. attribute:: box_centers
+
+        dim-by-nboxes :mod:`numpy` array of the centers of the boxes.
+
+    .. attribute:: box_parent_ids
+
+        :mod:`numpy` vector of parent box ids.
+
+    .. attribute:: box_child_ids
+
+        (2**dim)-by-nboxes :mod:`numpy` array of children box ids.
+
+    .. attribute:: box_levels
+
+        :mod:`numpy` vector of box levels.
+
+    .. automethod:: __init__
+
+    .. automethod:: copy
+
+    .. automethod:: get_leaf_flags
+
+    .. automethod:: leaf_boxes
+    """
+    box_centers: npt.NDArray
+    root_extent: npt.NDArray
+    box_parent_ids: npt.NDArray
+    box_child_ids: npt.NDArray
+    box_levels: npt.NDArray
+
+    def __post_init__(self):
+        lows = self.box_centers[:, 0] - 0.5*self.root_extent
+        highs = lows + self.root_extent
+        self.bounding_box = [lows, highs]
+        self.dim = self.box_centers.shape[0]
+
+    def copy(self):
+        return TreeOfBoxes(
+                box_centers=self.box_centers.copy(),
+                root_extent=self.root_extent.copy(),
+                box_parent_ids=self.box_parent_ids.copy(),
+                box_child_ids=self.box_child_ids.copy(),
+                box_levels=self.box_levels.copy())
```

Comment:
This smells like there's in-place modification somewhere.

Reply:
`copy` method has been removed